File: examples/python-functioncalling/extractwp.py
import requests
import json

model = "orca2"

systemprompt = "You will be given a text along with a prompt and a schema. You will have to extract the information requested in the prompt from the text and generate output in JSON observing the schema provided. If the schema shows a type of integer or number, you must only show a integer for that field. A string should always be a valid string. If a value is unknown, leave it empty. Output the JSON with extra spaces to ensure that it pretty prints."

schema = {
    "people": [
        {
            "name": {"type": "string", "description": "Name of the person"},
            "title": {"type": "string", "description": "Title of the person"},
        }
    ],
}

# Read the content from the file
words = []
with open("wp.txt") as f:
    maxwords = 2000
    count = 0
    lines = f.readlines()
    for line in lines:
        for word in line.split(" "):
            count += 1
            if count > maxwords:
                break
            words.append(word)
content = ' '.join(words)

# Use the text and schema to set the prompt
prompt = f"Review the source text and determine 10 the most important people to focus on. Then extract the name and title for those people. Output should be in JSON.\n\nSchema: {schema}\n\nSource Text:\n{content}"


# Make the actual request to the model
r = requests.post(
    "http://localhost:11434/api/generate",
    json={
        "model": model,
        "system": systemprompt,
        "prompt": prompt,
        "format": "json",
        "stream": False
    },
)

# Get the response as JSON.
j = json.loads(r.text)

# Return the result.
print(j["response"])

File: examples/typescript-functioncalling/extractwp.ts
import { Ollama } from "ollama-node";
import { readFile } from "fs/promises";

async function main() {

  const ollama = new Ollama();

  // Set the system prompt to prepare the model to receive a prompt and a schema and set some rules for the output.
  const systemprompt = `You will be given a text along with a prompt and a schema. You will have to extract the information requested in the prompt from the text and generate output in JSON observing the schema provided. If the schema shows a type of integer or number, you must only show a integer for that field. A string should always be a valid string. If a value is unknown, leave it empty. Output the JSON with extra spaces to ensure that it pretty prints.`

  const schema = {
    "people": [{
      "name": {
        "type": "string",
        "description": "Name of the person"
      },
      "title": {
        "type": "string",
        "description": "Title of the person"
      }
    }],
  }

  // Depending on the model chosen, you may be limited by the size of the context window, so limit the context to 2000 words.
  const textcontent = await readFile("./wp.txt", "utf-8").then((text) => text.split(" ").slice(0, 2000).join(" "));

  // Specific instructions for this task
  const prompt = `Review the source text and determine the 10 most important people to focus on. Then extract the name and title for those people. Output should be in JSON.\n\nSchema: \n${JSON.stringify(schema, null, 2)}\n\nSource Text:\n${textcontent}`

  await ollama.setModel("neural-chat");
  ollama.setSystemPrompt(systemprompt);

  // setJSONFormat is the equivalent of setting 'format: json' in the API
  ollama.setJSONFormat(true);
  await ollama.streamingGenerate(prompt, (word) => { process.stdout.write(word) })
}

main();